File: DevEnvironment/CANMessageGenerator/src/main/java/canmessagegenerator/core/CANRuleGenerator.java
package canmessagegenerator.core;

import base.core.RandomGenerator;
import base.interfaces.IRandomGenerator;
import java.util.ArrayList;
import java.util.List;

/**
 * The CANSignalRuleGenerator class implements a generator of CAN signal rules.
 */
public abstract class CANRuleGenerator {
    private static final String DEFAULT_NODE_NAME = "Vector__XXX";
    private static final String NODE_NAME_PREFIX = "Node";
    private static final int NUMBER_OF_NODES = 100;
    private static final List<String> transmittingNodeNames = createTransmittingNodeNames();

    private final IRandomGenerator randomGenerator = new RandomGenerator();

    /**
     * The CANRuleGenerator constructor.
     */
    protected CANRuleGenerator() {
    }

    /**
     * Generates a random generator.
     */
    protected IRandomGenerator randomGenerator() {
        return this.randomGenerator;
    }

    /**
     * Generates a new transmitting node name.
     */
    protected String generateTransmittingNodeName() {
        return this.generateTransmittingNodeName(transmittingNodeNames);
    }

    /**
     * Generates a new transmitting node name.
     */
    private String generateTransmittingNodeName(List<String> transmittingNodeNames) {
        int numberOfNames = transmittingNodeNames.size();
        int nameIndex = this.randomGenerator.nextInteger(0, numberOfNames - 1);

        String name = transmittingNodeNames.get(nameIndex);

        return name;
    }

    /**
     * Creates the transmitting node names.
     */
    private static List<String> createTransmittingNodeNames() {
        List<String> result = new ArrayList<>();
        result.add(DEFAULT_NODE_NAME);

        for (int i = 0; i < NUMBER_OF_NODES; ++i) {
            String nodeName = NODE_NAME_PREFIX + i;
            result.add(nodeName);
        }

        return result;
    }
}

File: DevEnvironmentCLion/CANMessageGenerator/core/CANRuleGenerator.cpp
#include "PreCompiled.h"
#include "CANRuleGenerator.h"

using namespace controllerareanetwork::messagegenerator;

const std::string CANRuleGenerator::defaultNodeName = "Vector__XXX";
const std::string CANRuleGenerator::nodeNamePrefix = "Name";
unsigned int CANRuleGenerator::numberOfNodes = 100;

/**
 * The CANRuleGenerator constructor.
 */
CANRuleGenerator::CANRuleGenerator()
{
}

/**
 * The CANRuleGenerator destructor.
 */
CANRuleGenerator::~CANRuleGenerator()
{
}

/**
 * Generates a new transmitting node name.
 */
const std::string& CANRuleGenerator::generateTransmittingNodeName()
{
    static std::unique_ptr<std::vector<std::string>> nodeNames(getTransmittingNodeNames());
    return generateTransmittingNodeName(*nodeNames);
}

/**
 * Generates a new transmitting node name.
 */
const std::string& CANRuleGenerator::generateTransmittingNodeName(const std::vector<std::string>& transmittingNodeNames)
{
    size_t numberOfNames = transmittingNodeNames.size();
    size_t nameIndex = m_randomGenerator.nextInteger(0, static_cast<int>(numberOfNames - 1));

    const std::string& name = transmittingNodeNames.at(nameIndex);

    return name;
}

/**
 * Gets the transmitting node names.
 */
std::vector<std::string>* CANRuleGenerator::getTransmittingNodeNames()
{
    std::unique_ptr<std::vector<std::string>> result = std::make_unique<std::vector<std::string>>();

    result->push_back(defaultNodeName);

    for (unsigned int i = 0; i < numberOfNodes; ++i) {
        std::string nodeName = nodeNamePrefix + std::to_string(i);
        result->push_back(nodeName);
    }

    return result.release();
}